File: go/find-bottom-left-tree-value.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 type TreeNodeLevel struct {
    Level int
    Node *TreeNode
}

func CreateTreeNodeLevel(node *TreeNode, level int) *TreeNodeLevel {
    var treeNodeLevel TreeNodeLevel
    treeNodeLevel.Level = level
    treeNodeLevel.Node = node
    return &treeNodeLevel
}

func findBottomLeftValue(root *TreeNode) int {
    stack := make([]*TreeNodeLevel, 0)
    stack = append(stack, CreateTreeNodeLevel(root, 0))
    var bottomLeftValue int
    level := -1
    for 0 < len(stack) {
        treeNodeLevel := stack[0]
        stack = stack[1:]
        if treeNodeLevel.Node == nil {
            continue
        }
        if level < treeNodeLevel.Level {
            level = treeNodeLevel.Level
            bottomLeftValue = treeNodeLevel.Node.Val
        }
        next_level := treeNodeLevel.Level + 1
        stack = append(stack, CreateTreeNodeLevel(treeNodeLevel.Node.Left, next_level))
        stack = append(stack, CreateTreeNodeLevel(treeNodeLevel.Node.Right, next_level))
    }
    return bottomLeftValue
}

File: python/find-bottom-left-tree-value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
        s = [(root, 0)]
        bottom_left_value = 0
        deeper_level = -1
        while s:
            node, level = s.pop()
            if node == None:
                continue
            if deeper_level < level:
                deeper_level = level
                bottom_left_value = node.val
            next_level = level + 1
            s.append((node.right, next_level))
            s.append((node.left, next_level))
        return bottom_left_value
